kubernetes/google-kubernetes-engine/bookie.yaml

```diff
@@ -17,120 +17,133 @@
-                args:
-                  - >
-                    bin/apply-config-from-env.py conf/bookkeeper.conf &&
-                    bin/bookkeeper shell metaformat --nonInteractive || true;
-                envFrom:
-                  - configMapRef:
-                        name: bookie-config
-
-            volumes:
-                # Mount local disks
-              - name: journal-disk
-                hostPath:
-                    path: /mnt/disks/ssd0
-              - name: ledgers-disk
-                hostPath:
-                    path: /mnt/disks/ssd1
-
+  name: bookie-hdd
+provisioner: kubernetes.io/gce-pd
+parameters:
+  type: pd-standard
+  zones: us-central1-a
 ---
-
-##
-## Define the Bookie headless service
-## In practice, in this case, it is only useful to have a view of
-## all the bookie pods that are present
-##
 apiVersion: v1
-kind: Service
+kind: ConfigMap
 metadata:
-    name: bookkeeper
-    labels:
-        app: pulsar
-        component: bookkeeper
+  name: bookie-config
+data:
+  PULSAR_MEM: "\" -Xms4g -Xmx4g -XX:MaxDirectMemorySize=4g\""
+  PULSAR_GC: "\" -XX:+UseG1GC \""
+  dbStorage_writeCacheMaxSizeMb: "1024"
+  dbStorage_readAheadCacheMaxSizeMb: "1024"
+  zkServers: zk-0.zookeeper,zk-1.zookeeper,zk-2.zookeeper
+  statsProviderClass: org.apache.bookkeeper.stats.PrometheusMetricsProvider
+---
+apiVersion: apps/v1beta1
+kind: StatefulSet
+metadata:
+  name: bookie
+  labels:
+    app: pulsar
+    component: bookkeeper
 spec:
-    ports:
-      - port: 3181
-        name: server
-    clusterIP: None
-    selector:
+  serviceName: "bookkeeper"
+  replicas: 3
+  selector:
+    matchLabels:
+      app: pulsar
+      component: bookkeeper
+  template:
+    metadata:
+      labels:
         app: pulsar
         component: bookkeeper
+        cluster: pulsar-gke
+      annotations:
+        pod.alpha.kubernetes.io/initialized: "true"
+        prometheus.io/scrape: "true"
+        prometheus.io/port: "8000"
+    spec:
+      terminationGracePeriodSeconds: 0
+      affinity:
+        podAntiAffinity:
+          preferredDuringSchedulingIgnoredDuringExecution:
+            - weight: 1
+              podAffinityTerm:
+                labelSelector:
+                  matchExpressions:
+                    - key: component
+                      operator: In
+                      values:
+                        - bookkeeper
+      containers:
+        - name: bookie
+          image: apachepulsar/pulsar:latest
+          command: ["sh", "-c"]
+          args:
+            - >
+              bin/apply-config-from-env.py conf/bookkeeper.conf &&
+              bin/apply-config-from-env.py conf/pulsar_env.sh &&
+              bin/pulsar bookie
+          ports:
+            - containerPort: 3181
+              hostPort: 3181
+              name: client
+          envFrom:
+            - configMapRef:
+                name: bookie-config
+          env:
+            - name: advertisedAddress
```

Comment:
@lucperkins This `advertisedAddress` was a setting we used for DeamonSet mode. With StatefulSet we should use the PodIP, which is also the hostname as we see from inside the container as the Bookie identifier. 

To do this, we don't need to do any fixing of the `advertisedAddress` but rather we should just set `useHostNameAsBookieID=true` in the config map above.